db/crud.go

```diff
@@ -1712,13 +1713,19 @@ func (db *DatabaseContext) assignSequence(ctx context.Context, docSequence uint6
 		// so we're allowing more 'recent sequences' on the doc (20) before attempting pruning
 		stableSequence := db.changeCache.GetStableSequence(doc.ID).Seq
 		count := 0
+		// we want to keep at least kMinRecentSequences recent sequences in the recent sequences list to reduce likelihood
+		// races between compaction of resent sequences and a coalesced DCP mutation resulting in skipped/abandoned sequences
+		maxToCompact := len(doc.RecentSequences) - kMinRecentSequences
 		for _, seq := range doc.RecentSequences {
 			// Only remove sequences if they are higher than a sequence that's been seen on the
 			// feed. This is valid across SG nodes (which could each have a different nextSequence),
 			// as the mutations that this node used to rev nextSequence will at some point be delivered
 			// to each node.
 			if seq < stableSequence {
 				count++
+				if count == maxToCompact {
+					break
+				}
 			} else {
 				break
 			}
```

Comment:
I wonder if this is easier to read, though you would want to do a check for len sequences outside of this function. You could create a function that _just_ prunes sequences to avoid longer logic and allow for early returns for len(doc.RecentSequences) < 5, etc.

Not set on this idea, just a thought on how to make it easier to read.

```suggestion
			if seq >= stableSequence {
				break
			}
			count++
			if count == maxCount {
				break
			}
```